### ISSUE TITLE
Element substitution with context

### ISSUE THREAD

**Issue by Akrugerus** [tefra/xsdata#685]
I'm working with an XML API that has a DTD defined for each endpoint but many endpoints share common types. What I've done so far is to use trang to convert the DTDs to XSDs and fed those into xsdata. Everything is working great except that each endpoint defines a request and response element that has different contents for each endpoint. This causes the request and response elements to flatten and only contain the child element for the first schema parsed.

So far I've considered unwrapping each response to its first unique element(immediately after the response element) or generating dataclasses for each endpoint individually. The first option requires more manual boilerplate and prevents access to the request element and the second option generates tons of unnecessary duplicate classes.

Since request and response are always children of a unique ENDPOINT_OUTPUT element, I'm wondering if there's an easy way to substitute the response elements with something like ENDPOINT_OUTPUT_RESPONSE which would make these elements unique but still allow the rest of the duplicate elements to flatten.

I've found that using a substitution like this:
```
<Config>
  <Substitutions>
    <Substitution type="class" search="Response" replace="Test"/>
  </Substitutions>
</Config>
```
allows me to manipulate the output class name but this obviously is not having the effect I'm looking for.

Any advice is appreciated.

**Comment by tefra** [comment]
Sorry I don't have any experience with DTDs

Maybe we can add the mapper for the DTD can you share the api dtd and schemas?

**Comment by Akrugerus** [comment]
The api I'm working with is the Qualys Vulnerability Management(VM) api. Here's a doc showing what the DTDs look like and providing links to fetch the dtds: https://www.qualys.com/docs/qualys-api-vmpc-xml-dtd-reference.pdf
At the moment I'm just storing a list of the DTD urls for the endpoints that I'm interacting with.

**Comment by Akrugerus** [comment]
The only place I've found comprehensive samples is in [another pdf](https://www.qualys.com/docs/qualys-api-vmpc-user-guide.pdf). I previously tried to parse the pdf for those samples but it got ugly quickly

**Comment by tefra** [comment]
HI @Akrugerus 

I just merged the first version of the dtd code generator, give it a try and let me know.

I added a test suite for xcbl in the [samples](https://github.com/tefra/xsdata-samples/tree/master/xcbl) repo, they offered a nice collection of schemas and samples and that seems to work fine.

I didn't get to test it against the qualys definitions, because it was a drag to try to extract them from that pdf. I would love to add it in the samples repo but it was taking a tone of time and I gave up when I found a couple of typos.

**Comment by Akrugerus** [comment]
Removing slots and kw-only args fixed that error but I'm still getting this one

```
./qualys_test.py::AssetGroupListTest::test_output Failed with Error: Unknown property IP_SET:IP
  File "/usr/local/Cellar/python@3.10/3.10.5/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/local/Cellar/python@3.10/3.10.5/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/usr/local/Cellar/python@3.10/3.10.5/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/placeholder/project-root/tests/qualys_test.py", line 53, in test_output
    r = qualys.list_asset_groups()
  File "/placeholder/project-root/package_root/qualys/asset_groups.py", line 69, in list_asset_groups
    output = parser.from_bytes(r.content, AssetGroupListOutput).response.asset_group_list
  File "/placeholder/project-root/.venv/lib/python3.10/site-packages/xsdata/formats/bindings.py", line 29, in from_bytes
    return self.parse(io.BytesIO(source), clazz)
  File "/placeholder/project-root/.venv/lib/python3.10/site-packages/xsdata/formats/dataclass/parsers/bases.py", line 53, in parse
    result = handler.parse(source)
  File "/placeholder/project-root/.venv/lib/python3.10/site-packages/xsdata/formats/dataclass/parsers/handlers/lxml.py", line 47, in parse
    return self.process_context(ctx)
  File "/placeholder/project-root/.venv/lib/python3.10/site-packages/xsdata/formats/dataclass/parsers/handlers/lxml.py", line 53, in process_context
    self.parser.start(
  File "/placeholder/project-root/.venv/lib/python3.10/site-packages/xsdata/formats/dataclass/parsers/bases.py", line 87, in start
    child = item.child(qname, attrs, ns_map, len(objects))
  File "/placeholder/project-root/.venv/lib/python3.10/site-packages/xsdata/formats/dataclass/parsers/nodes/element.py", line 344, in child
    raise ParserError(f"Unknown property {self.meta.qname}:{qname}")
xsdata.exceptions.ParserError: Unknown property IP_SET:IP
```

Here's a scrubbed snippet from the response that generated this error
```
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE ASSET_GROUP_LIST_OUTPUT SYSTEM "https://{platform_url}/api/2.0/fo/asset/group/asset_group_list_output.dtd">
<ASSET_GROUP_LIST_OUTPUT>
  <REQUEST>
    <DATETIME>2022-08-24T17:07:06Z</DATETIME>
    <USER_LOGIN>{my_username}</USER_LOGIN>
    <RESOURCE>https://{platform_url}/api/2.0/fo/asset/group/</RESOURCE>
    <PARAM_LIST>
      <PARAM>
        <KEY>action</KEY>
        <VALUE>list</VALUE>
      </PARAM>
      <PARAM>
        <KEY>echo_request</KEY>
        <VALUE>1</VALUE>
      </PARAM>
    </PARAM_LIST>
  </REQUEST>
  <RESPONSE>
    <DATETIME>2022-08-24T17:07:06Z</DATETIME>
    <ASSET_GROUP_LIST>
      <ASSET_GROUP>
        <ID>{Asset Group ID}</ID>
        <TITLE><![CDATA[{Asset Group Name}]]></TITLE>
        <NETWORK_ID>0</NETWORK_ID>
        <IP_SET>
          <IP>1.1.1.1</IP>
        </IP_SET>
      </ASSET_GROUP>
    </ASSET_GROUP_LIST>
  </RESPONSE>
</ASSET_GROUP_LIST_OUTPUT>
<!-- CONFIDENTIAL AND PROPRIETARY INFORMATION. Qualys provides the QualysGuard Service "As Is," without any warranty of any kind. Qualys makes no warranty that the information contained in this report is complete or error-free. Copyright 2022, Qualys, Inc. //--> 
```

**Comment by Akrugerus** [comment]
I've submitted a support ticket with Qualys to try to get those XML samples provided in a reasonable format, but I'm not holding my breath

**Comment by tefra** [comment]
Both errors I think were fixed in #703 

Give it a try @Akrugerus it's on master.


Please open new issues, I almost missed this one.
